### PR TITLE
[refactor](function) improve compoundPred optimization work with children is nullable

### PR DESCRIPTION
**Pull request by zhangstar333**
before this optimization work has limit, it's children must not nullable.

## Proposed changes

Issue Number: close #xxx

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by zhangstar333** [comment]
run buildall

File: be/src/vec/exprs/vcompound_pred.h [hunk 1]
```diff
@@ -52,86 +53,190 @@ class VCompoundPred : public VectorizedFnCall {
...
+        // false and NULL ----> 0
+        // true  and NULL ----> NULL
         if (_op == TExprOpcode::COMPOUND_AND) {
-            if (empty) {
-                // empty and any = empty, return lhs
+            //not null column: all data is false
+            //nullable column: null map all is not null
+            if ((lhs_all_false && !lhs_is_nullable) || (lhs_all_false && lhs_all_is_not_null)) {
+                // false and any = false, return lhs
                 *result_column_id = lhs_id;
```
Comment (BiteTheDDDDt) [inline be/src/vec/exprs/vcompound_pred.h:112]: If rhs is nullable, the return column need nullable too.

File: be/src/vec/exprs/vcompound_pred.h [hunk 2]
```diff
@@ -52,86 +53,190 @@ class VCompoundPred : public VectorizedFnCall {
...
             } else {
                 RETURN_IF_ERROR(get_rhs_colum());
 
-                if (full) {
-                    // full and any = any, return rhs
+                if ((lhs_all_true && !lhs_is_nullable) ||    //not null column
+                    (lhs_all_true && lhs_all_is_not_null)) { //nullable column
```
Comment (BiteTheDDDDt) [inline be/src/vec/exprs/vcompound_pred.h:117]: // need following whitespace

**Comment by zhangstar333** [comment]
run buildall

**Comment by zhangstar333** [comment]
run buildall

**Review by BiteTheDDDDt — DISMISSED** [review]

File: be/src/vec/exprs/vcompound_pred.h [hunk 3]
```diff
@@ -52,86 +54,196 @@ class VCompoundPred : public VectorizedFnCall {
...
-        bool empty_rhs = false;
+        uint8* __restrict rhs_data_column = nullptr;
+        uint8* __restrict rhs_null_map = nullptr;
+        bool rhs_is_nullable = false;
+        bool rhs_all_true = false;
+        bool rhs_all_false = false;
+        bool rhs_all_is_not_null = false;
+        bool result_is_nullable = _data_type->is_nullable();
 
         auto get_rhs_colum = [&]() {
```
Comment (HappenLee) [inline be/src/vec/exprs/vcompound_pred.h:89]: why not use same func to dispose the left column？

File: be/src/vec/exprs/vcompound_pred.h [hunk 4]
```diff
@@ -52,86 +54,196 @@ class VCompoundPred : public VectorizedFnCall {
...
             return VectorizedFnCall::execute(context, block, result_column_id);
         }
 
         int lhs_id = -1;
         int rhs_id = -1;
         RETURN_IF_ERROR(_children[0]->execute(context, block, &lhs_id));
-        ColumnPtr lhs_column =
-                block->get_by_position(lhs_id).column->convert_to_full_column_if_const();
-
+        //should deal with const column ?
```
Comment (HappenLee) [inline be/src/vec/exprs/vcompound_pred.h:64]: if there is const column？

**Comment by zhangstar333** [comment]
run buildall

File: be/src/vec/exprs/vcompound_pred.h (hunk 3 above)
Comment (zhangstar333) [reply be/src/vec/exprs/vcompound_pred.h:89]: it's need handle some local variables

File: be/src/vec/exprs/vcompound_pred.h (hunk 4 above)
Comment (zhangstar333) [reply be/src/vec/exprs/vcompound_pred.h:64]: have check the children are not const

**Review by BiteTheDDDDt — DISMISSED** [review]

File: be/src/vec/exprs/vcompound_pred.h (hunk 4 above)
Comment (HappenLee) [inline be/src/vec/exprs/vcompound_pred.h:64]: del the line

File: be/src/vec/exprs/vcompound_pred.h [hunk 5]
```diff
@@ -52,86 +54,196 @@ class VCompoundPred : public VectorizedFnCall {
...
+            *result_column_id = block->columns();
+            block->insert({std::move(result_column), _data_type, _expr_name});
+        };
+
+        auto vector_vector_null = [&]<bool is_and_op>() {
+            auto col_res = ColumnUInt8::create(size);
+            auto col_nulls = ColumnUInt8::create(size);
+            auto* __restrict res_datas = assert_cast<ColumnUInt8*>(col_res)->get_data().data();
+            auto* __restrict res_nulls = assert_cast<ColumnUInt8*>(col_nulls)->get_data().data();
+            ColumnPtr temp_null_map = nullptr;
```
Comment (HappenLee) [inline be/src/vec/exprs/vcompound_pred.h:120]: use a function do the logic

**Comment by zhangstar333** [comment]
run buildall

**Review by HappenLee — APPROVED** [review]
LGTM

**Review by BiteTheDDDDt — APPROVED** [review]

**Pull request merged**